### DISCUSSION TITLE
Native library binding - Mapbox

### DISCUSSION THREAD

**Discussion by steveres** [dotnet/maui#578] [General] [answered]
[Mapbox ](https://www.mapbox.com/) is the best mapping library for outdoor adventure consumer-facing apps (think cycling, hiking, running). It uses vector-based tiles, that are faster to render, better for offline use cases, and have greater resolution than raster-based tiles used by Google maps. The Mapbox library is quite complex so creating bindings is a daunting task. In addition, they update their libraries often, so keeping up (dev + test) would be very time consuming. There has been [one attempt](https://github.com/NAXAM/mapbox-xamarin-forms) at this, which stalled. 

Is it possible to leverage a native 3rd party UI-intensive control w/o developing and maintaining bindings? Pretty sure the answer is "no" for now, but wondering how a future version of Xamarin or .NET MAUI can solve this problem.

Without this ability, the gold-standard in mapping is out of reach for Xamarin-based apps. As the author of [Bicycle Route Navigator](https://apps.apple.com/us/app/bicycle-route-navigator/id1320473922), we're at an inflection point on potentially rebuilding from Xamarin to native iOS and Android apps to leverage Mapbox.

**Comment by StarChick971** [comment]
Any news concerning Mapbox? or even map solution regarding .NET MAUI? I saw some news about embedding a map in their roadmap (starting with Google Maps I guess) but that's was several months ago (I remember they said around december 2021 or Q1 2022)...

**Comment by czmirek** [comment]
I just want to write here that you can use `BlazorWebView` for running MapBox in MAUI. I managed to do that and it's working well.

Even if your whole app is written in XAML, you can use `BlazorWebView` just to render & control the page where only the map is shown. It runs on the official MapBox JS API and you interact with it with JSInterop.

Performance-wise it's probably going to be a bit slower than an optimized native solution but I can confirm it's definitely a working solution.

**Reply by StarChick971** [reply]
Thank you for the reply.
Is it really needed to use a Blazor MAUI app? I would like to make it as easy to maintain as possible but I keep your suggestion in mind.
Also, I wish to get best performance to save battery, because optimized apps have less power consumption, and I am building an outdoor app for extreme weather conditions especially cold in mountains 😉

**Reply by czmirek** [reply]
The problem is that for MapBox to work natively you need to create bindings for the official Android library (AAR or JAR).

This process right now is quite difficult and not well documented (see the resources below) and you have to spend at least few days learning how to create these bindings. It's not a straightforward process because of differences between Java and C#.

...but if you learn it you give yourself freedom to use any Android library in your .NET MAUI project.

...or wait for some good soul that will do it for us and publish it :)

I personally gave up trying to create the bindings for MapBox AAR and switched to Google maps for now.

Beware that you probably need to create bindings not only for the MapBox AAR but also for its dependencies (which may or may not exist as NuGets).

Resources for learning:
https://learn.microsoft.com/en-us/xamarin/android/platform/binding-java-library/
https://gist.github.com/JonDouglas/dda6d8ace7d071b0e8cb

**Comment by madmapper** [comment]
Mapbox is great - but isn't the only game in town.  ESRI's ArcGIS Runtime SDK supports most major platforms including XF and they have a preview out for MAUI:
https://community.esri.com/t5/arcgis-runtime-sdks-blog/announcing-preview-1-of-net-maui-support-with/ba-p/1160505

We've been building with their SDK for years and they do support vector tile basemaps, offline data collection/sync, routing, etc... In the past ESRI hasn't really had a very great understanding of the commercial app dev market, but they are improving and have some competitively priced developer plans that pencil out quite competitively with MapBox's offerings

**Comment by tuyen-vuduc** [comment, marked as answer]
Hi forks,

Please check out my MAUI component for Mapbox.

https://www.nuget.org/packages/Mapbox.Maui

Regards.

**Reply by StarChick971** [reply]
Thank you so much for providing this new package! 🤩
However, I didn't manage to display a basic map following the github readme... Any tutorial ?
I am using the latest version of Visual Studio Preview 17.7 then I tried either with .NET7 LTS and .NET8 Preview but doesn't work...

**Reply by tuyen-vuduc** [reply]
I am preparing a blog post and also revise the readme file :D .

**Reply by tuyen-vuduc** [reply]
And please follow REAMDE file strictly and slowly :D . It requires some edits on `csproj` and also a system wide `gradle.properties` file.

**Reply by tuyen-vuduc** [reply]
The other thing: MacOS will be the preferred OS to run :D.

**Reply by tuyen-vuduc** [reply]
To run iOS, you need `MapboxMauiQs.props` file with the below content.
```
<?xml version="1.0" encoding="utf-8"?>
<Project>
	<PropertyGroup>
		<MAPBOX_DOWNLOADS_TOKEN>YOUR_MAPBOX_DOWNLOADS_TOKEN</MAPBOX_DOWNLOADS_TOKEN>
	</PropertyGroup>
</Project>
```

This file will be in the same folder of `MapboxMauiQs.csproj` file

**Reply by StarChick971** [reply]
Actually I am more focusing on Android at the moment, but iOS is the next step once it works on Android. 😅

**Reply by tuyen-vuduc** [reply]
Here is the guide to integrate Mapbox Android SDK to your app. 

https://tuyen-vuduc.tech/how-to-use-mapbox-for-your-dotnet-android-app

**Reply by StarChick971** [reply]
I meant, I created a .NET MAUI app but I only own an Android device for testing 😅
Btw, following your MAUI tutorial I can't see any  **gradle** folder or file within a fresh new .NET MAUI project created by Visual Studio 🤔

**Reply by saterry** [reply]
So the .gradle folder is under C:\Users\<user>\.gradle  Hopefully this saves someone from a headache.

Now I'm trying to figure out why my build is getting Could not resolve ... Received status code 401 from server: Unauthorized
I am setting the public and private keys.  (As far as I know)

**Reply by tuyen-vuduc** [reply]
@StarChick971 , @saterry 

Have you figured out how to come over it?

Please follow the guide slowly and strictly.

It's a bit complicated than you might think of.

**Reply by kkouer** [reply]
@tuyen-vuduc Hello, I also used the Mapbox.maui control you developed and ran into problems with the error code GDL001 Gradlew build failed D:\MAUIServer\MauiDataCenter\MauiDataCenter\obj\xgradle\gradlew.bat, I noticed that your development environment is on macOS, and my test platform is windows, I wonder if you can provide some help information, thank you

**Reply by DincaAndrei2410** [reply]
Maybe you omit [this](https://tuyen-vuduc.tech/how-to-use-mapbox-in-your-dotnet-maui-app?source=more_series_bottom_blogs#heading-1-config-your-mapboxdownloadstoken) step? 

Without adding gradle.properties inside .gradle home directory would not work. Also if you did that and still not working, inside .gradle/daemon/7.6 you should find some log files with more info that might help pinpoint the problem.

**Reply by kkouer** [reply]
Yes I did this by manually adding MAPBOX_DOWNLOADS_TOKEN to the.gradle.properties file. If added in the console there will be a null character after each character, thank you for your help

**Comment by DevAayushRawat** [comment]
I am tyring to implement the solution as you did but **getting this error** 
Gradlew build failed C:\Users\QSITA034\source\repos\MapboxMaui\MapboxMaui\obj\xgradle\gradlew.bat --> C:\Users\QSITA034\source\repos\MapboxMaui\MapboxMaui\obj\xgradle	MapboxMaui	C:\Users\QSITA034\.nuget\packages\dependencies.gradle\7.6.0.1\buildTransitive\Dependencies.Gradle.targets	29

**Reply by tuyen-vuduc** [reply]
This is the issue on Windows only. :D . There will be patch very soon. Please move the issue to the qs repo or the binding library repo.

**Reply by DevAayushRawat** [reply]
Will it work on mac?

**Reply by tuyen-vuduc** [reply]
Yes, it should. Many tried and were successful. Windows is a bit different from MacOS, hence requires additional work.

**Reply by DevAayushRawat** [reply]
Unfortunately it is not working on mac also!! Getting the same error on mac!

**Reply by tuyen-vuduc** [reply]
Have you followed the steps in [my get-started guide](https://tuyen-vuduc.tech/how-to-use-mapbox-in-your-dotnet-maui-app) to configure things correctly? 

You may need to clean then rebuild as well :D .

**Reply by DevAayushRawat** [reply]
Yes i did the same as you've told, but it's not working!!

**Reply by tuyen-vuduc** [reply]
I will check that out then let you know

**Reply by riburajan** [reply]
I am also getting the same error. Any solution for this? 

I am using MAUI(with visual studio 2022).

**Reply by nmpasdanis** [reply]
I am also getting the same error. Visual studio 2022.
Any solution ?

**Reply by tuyen-vuduc** [reply]
I will release the patch during the weekend. Please bear with me.

**Reply by riburajan** [reply]
Hi @tuyen-vuduc ,
Sure, we will wait for your reply.
Thanks

**Reply by nmpasdanis** [reply]
@tuyen-vuduc, Of course!!!

**Reply by riburajan** [reply]
Hi @tuyen-vuduc, 
Hope you spend time for our build errors in MAUI app.

Thanks

**Reply by tuyen-vuduc** [reply]
Can you try upgrading Mapbox Android to latest version?

https://www.nuget.org/packages/Com.Mapbox.Maps.Android/10.16.1.1

**Reply by riburajan** [reply]
@tuyen-vuduc , 
Is this update is for MAUI Apps right(both iOS and Android) ? 
I installed the latest version of Mapbox.Maui (version 10.11.1.1). But still the Build errors are exists. pls see the screen shot.

<img width="766" alt="2023-11-07 07_39_48-MapBox Map Maui - Microsoft Visual Studio" src="https://github.com/dotnet/maui/assets/94977917/d3da5742-e65d-412a-850e-27874519aced">

**Reply by tuyen-vuduc** [reply]
Plz follow the readme carefully and install the Android package as well to ensure the latest Android version is installed.

**Reply by riburajan** [reply]
Hi,
sure, will do that update the status here. 

for MAUI app, do we need to install Com.Mapbox.Maps.Android and Mapbox.Maui ?  Or only Mapbox.Maui is enough.?

**Reply by tuyen-vuduc** [reply]
To have the Android patch I just published, you must install the Android package as well.

**Reply by riburajan** [reply]
Hi @tuyen-vuduc ,

As you mentioned in the step by step document - 

**Put that token ~/.gradle/gradle.properties from the terminal (Replace YOUR_MAPBOX_DOWNLOADS_TOKEN with yours)**

 we are not able to add this line, Because before installing the package(**Mapbox.Maui 10.11.1.1** ) , these gridle folder is not present.

So then we try to install the package : **Mapbox.Maui 10.11.1.1**  this below error is coming: 

<img width="927" alt="image" src="https://github.com/dotnet/maui/assets/94977917/bbb42fff-5984-47d9-8bcd-f7fa665f34d9">

After this, tried to install : **Com.Mapbox.Maps.Android** . Then getting this error :

<img width="617" alt="image" src="https://github.com/dotnet/maui/assets/94977917/e710d7df-673f-46e3-bbc0-c18b096946ea">


Can you please add MAUI project in git reop using Visual Studio 2022? so we can refer that project. Please share the git repo link as well.  

Thanks

**Reply by tuyen-vuduc** [reply]
@riburajan

 VS2022 keeps showing the error because it builds the project right when we open it (create new one), even when we add new nuget packages.
 
Please FOLLOW the guide regardless of those errors.

To put the token in `~/.gradle/gradle.properties` first, here are the steps:
1. Create folder `.gradle` under your Windows user account if it doesn't exist
2. Create text file `gralde.properties` under that new folder if it doesn't exist
3. Ensure the access token is there as noted

On Windows, you can use `Git Bash` to edit  `~/.gradle/gradle.properties` using `vim` easily without creating `.gradle` folder prior to editing the file.

Regarding package version mismatch, it's VS2022 limitation, please install the highest version showing in the error of the dependency package before installing our desired package.

Cheers.

**Reply by riburajan** [reply]
Hi @tuyen-vuduc ,
We tried the same what you explaind in the steps. But still we are not able to install the package : Mapbox.Maui 10.11.1.1.

**After installing the package through NuGet Package Manager, getting this error :** 

<img width="607" alt="image" src="https://github.com/dotnet/maui/assets/94977917/b34884f6-e2e0-4767-a8f3-0fc91257d9b8">


What we can do for a success build?? please reply......

Thanks

**Reply by tuyen-vuduc** [reply]
@riburajan  Plz follow the steps carefully. Otherwise, please contact me via email. A direct call will help.

**Reply by riburajan** [reply]
Hi @tuyen-vuduc ,
I already sent an email to the Id: [hi@tuyen-vuduc.tech], Please check.

**Reply by XamTechie** [reply]
@riburajan Is it worked for you? I am getting the same issues

**Reply by DincaAndrei2410** [reply]
I've managed to fix a similar error by setting JAVA_HOME env variable to path corresponding to jdk-19. There is a compatibility [matrix](https://docs.gradle.org/current/userguide/compatibility.html) between gradle version and Java versions and in my case I had Java 21 which is not compatible with gradle 7.6. If this doesn't fix the issue, you can also check inside .gradle folder, there should be a folder daemon which inside has logs for different versions. In those logs you can find more info

**Reply by Saurus119** [reply]
**If your Android SDK is installed on a drive other than C:\, Android/Gradle may fail when trying to automatically install required components.**

`\dependencies.gradle\8.6.0.3\buildTransitive\Dependencies.Gradle.targets(30,3): error GDL001: Gradlew build failed 
obj\xgradle\gradlew.bat 
`

Detailed error output:
```
       Checking the license for package Android SDK Build-Tools 34 in E:\NetSDK\Android\android-sdk\licenses 
      8>License for package Android SDK Build-Tools 34 accepted. 8>Preparing "Install Android SDK Build-Tools 34 v.34.0.0". 
      8>Warning: Failed to read or create install properties file. 8> 8>FAILURE: Build failed with an exception. 
      8>* What went wrong: 8>Could not determine the dependencies of task ':gradle_util:compileDebugJavaWithJavac'. 
      8> Failed to install the following SDK components: 
      8> build-tools;34.0.0 Android SDK Build-Tools 34 
      **8> !!!!!!!!!!!!!!!! The SDK directory is not writable (E:\NetSDK\Android\android-sdk)** !!!!!!!!!!
      8> * Try: 
      8>> Run with --stacktrace option to get the stack trace. 
      8> Run with --info or --debug option to get more log output. 
      8> Run with --scan to get full insights. 
      8> Get more help at https://help.gradle.org. 8> 8>BUILD FAILED in 1s
```
**Resolution**
Give your user/system Full Control over the folder where the Android SDK is located — or move the SDK to a normal writeable location such as: **C:\Android\android-sdk**

- **Permission approach** -> Right click folder ->Security -> Give Full Control to your user and system account..
- **Move folder under C:\Android** -> Should be solution as well, but I didn´t test it as windows should be able write there without issue?

Without full permission on the SDK folder, Android SDK Build Tools (e.g. 34.0.0) cannot be installed, and the build will always fail.

**Comment by shahrukh339** [comment]
Hi @tuyen-vuduc  I also facing this issue 
![Screenshot 2024-07-03 at 11 12 52 AM](https://github.com/dotnet/maui/assets/80701098/2212fbbb-521c-4ddf-a4c9-3eb741ac5106)

**Reply by shahrukh339** [reply]
how can i solve this issue

**Reply by tuyen-vuduc** [reply]
Did you follow the steps outlined in the get started guide slowly and carefully? We have to do the way written in the official documents from Mapbox to align with their license of lib usage.

**Reply by DincaAndrei2410** [reply]
Inside (Home)/.gradle/ have you added a gradle.properties file with MAPBOX_DOWNLOADS_TOKEN=(your mapbox token)?

If so, you should find inside .gradle/daemon some folders named after gradle version being used. With latest version of mapbox, you should see a folder 8.6. In that folder, you can find some logs. Could you check for errors and paste them here?

**Reply by shahrukh339** [reply]
@DincaAndrei2410  thanks for your support.Are you talking about this ?

<img width="1440" alt="image" src="https://github.com/dotnet/maui/assets/80701098/035194cf-5f04-4381-841c-286f1da2af59">

<img width="1440" alt="image" src="https://github.com/dotnet/maui/assets/80701098/e84af35d-5085-49c7-a9b3-b62a013c509d">

**Reply by DincaAndrei2410** [reply]
On your system, in your home directory, you should find a .gradle folder and in that folder, either you already have a gradle.properties file or you create one. Either way, there you should put the mapbox token.

You can find a detailed explanation [here](https://tuyen-vuduc.tech/how-to-use-mapbox-in-your-dotnet-maui-app?source=more_series_bottom_blogs#heading-1-config-your-mapboxdownloadstoken)

**Reply by shahrukh339** [reply]
Hi @DincaAndrei2410 

I follow all the steps but getting other type of error 


![image](https://github.com/dotnet/maui/assets/80701098/6a1c0400-e209-4295-8469-6e757376359d)
![Uploading image.jpg…]()

**Reply by shahrukh339** [reply]
My gradle.properties look like this 
![Uploading image.jpg…]()

**Reply by tuyen-vuduc** [reply]
You didn't follow [the guide](https://tuyen-vuduc.tech/how-to-use-mapbox-in-your-dotnet-maui-app?source=more_series_bottom_blogs#heading-1-config-your-mapboxdownloadstoken).

> DOWNLOAD_TOKEN is different from ACCESS_TOKEN. Each has its own purpose and permissions.

Read it again slowly and carefully.

Plz check out Mapbox document as well for further info.

**Reply by tuyen-vuduc** [reply]
In the post to [get started for Mapbox on .NET Android](https://tuyen-vuduc.tech/how-to-use-mapbox-for-your-dotnet-android-app#heading-1-config-your-mapboxdownloadstoken), you will see the detailed steps of getting a download token. Plz check it out.

**Comment by shahrukh339** [comment]
Let me check again

**Comment by BurhanMunir** [comment]
@tuyen-vuduc Hi, I am implementing mapbox in .net maui, and have followed each step from the blog, adding accesstoken in my gradle properties,

and verify that access token exists there,

I have added .targets and .template files.


and I am facing this error

<img width="1044" alt="Screenshot 2025-04-12 at 11 31 26 PM" src="https://github.com/user-attachments/assets/c3bf6c6d-b08e-4a68-a62c-26fc9ef82a8c" />


please guide me in this scenario,

and I have also sent you the email, mentioning all the details and screen shots.

**Reply by tuyen-vuduc** [reply]
@BurhanMunir Please check out [my guide](https://tuyen-vuduc.tech/how-to-use-mapbox-in-your-dotnet-maui-app?source=more_series_bottom_blogs#heading-1-config-your-mapboxdownloadstoken) again carefully to ensure configure `MAPBOX_DOWNLOAD_TOKEN ` in `gradle.properties` file in folder `.gradle` in your home folder `~/.gradle/gradle.properties`

Happy coding!!!

**Reply by BurhanMunir** [reply]
@tuyen-vuduc definitely I follow your guide , again and again, And I am facing the above mentioned issue in screen shot, gradle issue,

Yes I have make sure MAPBOX_DOWNLOAD_TOKEN  in gradle.properties exists, but still facing issue,

I need the guidance resolving this issue, how to tackle it, please guide me if I am missing something in that sense.

**Reply by BurhanMunir** [reply]
![Screenshot 2025-04-16 145606](https://github.com/user-attachments/assets/9621043c-25db-47d1-a8d3-bf08c815a34d)
@tuyen-vuduc Now I am facing this issue, I have installed jdk-19 and latest version of gradle,

but unable to resolve this, 

and I will agree with @czmirek , we need to implement native library bindings, 

is it correct. There are a lot of gradle issues reported about mapbox, can you please reply or provide a sample project working fine,

yes, again I tried my best to follow your guide step by step.

so please make it clear, if mapbox working for .net maui or not stable yet.

Thanks

**Reply by tuyen-vuduc** [reply]
@BurhanMunir Thanks for using my library. For your question, the library is already in use for production app. If you find any issues, please help a hand to fix.

Regarding your above issue, I will need to check but my client didn't report a similar issue.

**Reply by BurhanMunir** [reply]
@tuyen-vuduc That's great, ok I have metnioned the steps I am following, can you please look into if I am missing something?

**Reply by BurhanMunir** [reply]
@tuyen-vuduc Hi, I resolved the issues, mentionded above. Here are the steps I followed and the things I was just missing.

1. First thing I was missing token with Downloads:Read permission, I was using public token.

2. Second I just missing the reference to MapboxMauiQs.targets file in my csproj file so I just reference it as

<img width="1085" alt="Screenshot 2025-04-20 at 3 04 25 PM" src="https://github.com/user-attachments/assets/5a67b0fe-e0ef-4421-a886-4c4f5ce03aeb" />


3. And Last one, on my windows I have added MAPBOX_DOWNLOADS_TOKEN in my gradle.properties but my project from visual 

studio was not picking the token so I was facing gradle or file not found issues, so I added MAPBOX_DOWNLOADS_TOKEN as my 

environment variable in my windows, and that's working fine,

@tuyen-vuduc Really thanks my friends for responding my messages and off course for the cool stuff you just delivered as mapbox 

integration in .net maui.


@riburajan @shahrukh339

**Reply by tuyen-vuduc** [reply]
@BurhanMunir Great to hear that.

**Comment by rajeshangappan** [comment]
Any plan to upgrade .NET 10?

**Reply by tuyen-vuduc** [reply]
@rajeshangappan  Thanks for your interest. I don't have a plan yet as I am no longer with .NET MAUI atm.

You can do the upgrade then submit a PR.

However, are there any issues ?

**Reply by riburajan** [reply]
@rajeshangappan, integrating Mapbox with MAUI is a little tricky. You may encounter several unexpected behaviors in Visual Studio—I faced similar issues myself. However, @tuyen-vuduc  helped me a lot. I connected with him last year, and he walked me through the setup on his computer. Unfortunately, we later had to drop that project.

@tuyen-vuduc, thank you for your support.  Hope you doing well.
